tests/binning_test.py

```diff
@@ -547,3 +547,60 @@ def test_variable_bin_equivalent_to_bin_of_data_array_with_counts():
     coord = sc.array(dims=['row'], values=[2, 2, 1, 3])
     da = sc.DataArray(data, coords={'x': coord})
     assert sc.identical(coord.bin(x=3), da.bin(x=3))
+
+
+def test_group_many_to_many_uses_optimized_codepath():
+    size = 512
+    table = sc.data.table_xyz(int(1e3))
+    da = table.bin(x=size, y=size)
+    xcoarse = sc.arange('x', size) // 2
+    ycoarse = sc.arange('y', size) // 2
+    da.coords['xcoarse'] = xcoarse
+    da.coords['ycoarse'] = ycoarse
+    # The old "standard" implementation uses a massive amount of memory, crashing even
+    # with 256 GByte of RAM.
+    assert da.group('xcoarse', 'ycoarse').dims == ('xcoarse', 'ycoarse')
+
+
+def test_group_many_to_many_with_extra_inner_dim_uses_optimized_codepath():
+    size = 512
+    table = sc.data.table_xyz(int(1e3))
+    da = table.bin(x=size, y=size, z=3)
+    xcoarse = sc.arange('x', size) // 2
+    ycoarse = sc.arange('y', size) // 2
+    da.coords['xcoarse'] = xcoarse
+    da.coords['ycoarse'] = ycoarse
+    # The old "standard" implementation uses a massive amount of memory, crashing even
+    # with 256 GByte of RAM.
+    assert da.group('xcoarse', 'ycoarse').dims == ('z', 'xcoarse', 'ycoarse')
+
+
+def test_group_many_to_many_with_extra_outer_dim_uses_optimized_codepath():
+    size = 512
+    table = sc.data.table_xyz(int(1e3))
+    da = table.bin(z=3, x=size, y=size)
+    xcoarse = sc.arange('x', size) // 2
+    ycoarse = sc.arange('y', size) // 2
+    da.coords['xcoarse'] = xcoarse
+    da.coords['ycoarse'] = ycoarse
+    # The old "standard" implementation uses a massive amount of memory, crashing even
+    # with 256 GByte of RAM.
+    assert da.group('xcoarse', 'ycoarse').dims == ('z', 'xcoarse', 'ycoarse')
+
+
+def test_group_many_and_erase():
+    from scipp import binning
+    binning.make_binned
+    size = 512
+    table = sc.data.table_xyz(int(1e3))
+    da = table.bin(z=3, x=size, y=size)
+    xcoarse = sc.arange('x', size) // 2
+    ycoarse = sc.arange('y', size) // 2
+    da.coords['xcoarse'] = xcoarse
+    da.coords['ycoarse'] = ycoarse
+    result = binning.make_binned(
+        da,
+        edges=[],
+        groups=[sc.arange('xcoarse', size // 2, dtype=xcoarse.dtype)],
+        erase=['x', 'y'])
+    assert result.dims == ('z', 'xcoarse')
```

Comment:
We need tests that ensure that the actual bin contents are correct and not just the dims. Is there a way to write a generic test that goes through all possible dim orderings? I have a hard time to reason about all cases.

Reply:
In the other test file I compared to `groupby.concat`, but as I am planning to deprecate that that also needs to change...

It is possible to bin directly into the result coords (from a table) and compare to that, but we might be limited to comparing the result of histogramming as the order within bins is not necessarily the same.

Reply:
Can we use the old, generic implementation to compare with the new one? Or would this be too much of a whitebox test?

Reply:
We could, but probably not too useful? Furthermore, we might remove support for some of this from the old implementation (to simplify it), and then all our tests are useless? I think in the end comparing against the histogram may be correct? We just need to find a way to make it sensitive to errors, i.e., somehow avoid event weights that add up to the same histogram?